File: src/main/java/org/lanternpowered/server/service/permission/UserSubject.java
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.service.permission;

import org.lanternpowered.server.config.user.OpsEntry;
import org.lanternpowered.server.game.Lantern;
import org.lanternpowered.server.profile.LanternGameProfile;
import org.lanternpowered.server.service.permission.base.LanternSubject;
import org.lanternpowered.server.service.permission.base.SingleParentMemorySubjectData;
import org.spongepowered.api.Sponge;
import org.spongepowered.api.command.CommandSource;
import org.spongepowered.api.profile.GameProfile;
import org.spongepowered.api.service.context.Context;
import org.spongepowered.api.service.permission.MemorySubjectData;
import org.spongepowered.api.service.permission.PermissionService;
import org.spongepowered.api.service.permission.Subject;
import org.spongepowered.api.service.permission.SubjectCollection;
import org.spongepowered.api.service.permission.SubjectReference;
import org.spongepowered.api.util.Tristate;

import java.util.Optional;
import java.util.Set;

/**
 * An implementation of vanilla minecraft's 4 op groups.
 */
final class UserSubject extends LanternSubject {

    private final GameProfile player;
    private final MemorySubjectData data;
    private final UserCollection collection;

    UserSubject(GameProfile player, UserCollection users) {
        this.player = player;
        this.collection = users;
        this.data = new SingleParentMemorySubjectData(this) {

            @Override
            public SubjectReference getParent() {
                int opLevel = getOpLevel();
                return opLevel == 0 ? null : users.getService().getGroupForOpLevel(opLevel).asSubjectReference();
            }

            @Override
            public void setParent(SubjectReference parent) {
                int opLevel;
                if (parent == null) {
                    opLevel = 0;
                } else {
                    final Subject subject = parent.resolve().join();
                    if (!(subject instanceof OpLevelCollection.OpLevelSubject)) {
                        return;
                    }
                    opLevel = ((OpLevelCollection.OpLevelSubject) subject).getOpLevel();
                }
                if (opLevel > 0) {
                    Lantern.getGame().getOpsConfig().addEntry(new OpsEntry(((LanternGameProfile) player).withoutProperties(), opLevel));
                } else {
                    Lantern.getGame().getOpsConfig().removeEntry(player.getUniqueId());
                }
            }
        };
    }

    @Override
    public String getIdentifier() {
        return this.player.getUniqueId().toString();
    }

    @Override
    public Optional<String> getFriendlyIdentifier() {
        return this.player.getName();
    }

    @SuppressWarnings("unchecked")
    @Override
    public Optional<CommandSource> getCommandSource() {
        if (Sponge.isServerAvailable()) {
            return (Optional) Lantern.getServer().getPlayer(this.player.getUniqueId());
        }
        return Optional.empty();
    }

    public int getOpLevel() {
        return Lantern.getGame().getOpsConfig().getEntryByUUID(this.player.getUniqueId()).map(OpsEntry::getOpLevel).orElse(0);
    }

    @Override
    public SubjectCollection getContainingCollection() {
        return this.collection;
    }

    @Override
    public MemorySubjectData getSubjectData() {
        return this.data;
    }

    @Override
    public PermissionService getService() {
        return this.collection.getService();
    }

    @Override
    public Tristate getPermissionValue(Set<Context> contexts, String permission) {
        Tristate ret = super.getPermissionValue(contexts, permission);
        if (ret == Tristate.UNDEFINED) {
            ret = getDataPermissionValue(this.collection.getDefaults().getSubjectData(), permission);
        }
        if (ret == Tristate.UNDEFINED) {
            ret = getDataPermissionValue(this.collection.getService().getDefaults().getSubjectData(), permission);
        }
        if (ret == Tristate.UNDEFINED && getOpLevel() >= Lantern.getGame().getGlobalConfig().getDefaultOpPermissionLevel()) {
            ret = Tristate.TRUE;
        }
        return ret;
    }

    @Override
    public Optional<String> getOption(Set<Context> contexts, String option) {
        Optional<String> ret = super.getOption(contexts, option);
        if (!ret.isPresent()) {
            ret = getDataOptionValue(this.collection.getDefaults().getSubjectData(), option);
        }
        if (!ret.isPresent()) {
            ret = getDataOptionValue(this.collection.getService().getDefaults().getSubjectData(), option);
        }
        return ret;
    }
}

File: src/main/kotlin/org/lanternpowered/server/service/permission/UserSubject.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.service.permission

import org.lanternpowered.api.profile.copyWithoutProperties
import org.lanternpowered.server.config.user.OpsEntry
import org.lanternpowered.server.game.Lantern
import org.lanternpowered.server.profile.LanternGameProfile
import org.lanternpowered.server.service.permission.OpLevelCollection.OpLevelSubject
import org.spongepowered.api.profile.GameProfile
import org.spongepowered.api.service.context.Context
import org.spongepowered.api.service.permission.MemorySubjectData
import org.spongepowered.api.service.permission.PermissionService
import org.spongepowered.api.service.permission.SubjectCollection
import org.spongepowered.api.service.permission.SubjectReference
import org.spongepowered.api.util.Tristate
import java.util.Optional
import java.util.UUID

/**
 * An implementation of vanilla minecraft's 4 op groups.
 */
internal class UserSubject(
        private val player: GameProfile,
        private val collection: UserCollection
) : LanternSubject() {

    private val data: MemorySubjectData

    init {
        this.data = object : SingleParentMemorySubjectData(this) {

            override var parent: SubjectReference?
                get() {
                    val opLevel = opLevel
                    return if (opLevel == 0) null else collection.service.getGroupForOpLevel(opLevel).asSubjectReference()
                }
                set(parent) {
                    val opLevel = if (parent == null) {
                        0
                    } else {
                        val subject = parent.resolve().join() as? OpLevelSubject ?: return
                        subject.opLevel
                    }
                    if (opLevel > 0) {
                        Lantern.getGame().opsConfig.addEntry(OpsEntry(player.copyWithoutProperties(), opLevel))
                    } else {
                        Lantern.getGame().opsConfig.removeEntry(player.uniqueId)
                    }
                }
        }
    }

    val uniqueId: UUID
        get() = this.player.uniqueId

    override fun getIdentifier(): String = this.player.uniqueId.toString()
    override fun getFriendlyIdentifier(): Optional<String> = this.player.name
    override fun getContainingCollection(): SubjectCollection = this.collection
    override fun getSubjectData(): MemorySubjectData = this.data

    val opLevel: Int
        get() = Lantern.getGame().opsConfig.getEntryByUUID(this.player.uniqueId).map { obj -> obj.opLevel }.orElse(0)

    override val service: PermissionService
        get() = this.collection.service

    override fun getPermissionValue(contexts: Set<Context>, permission: String): Tristate {
        var result = super.getPermissionValue(contexts, permission)
        if (result == Tristate.UNDEFINED)
            result = this.getDataPermissionValue(this.collection.defaults.subjectData, permission)
        if (result == Tristate.UNDEFINED)
            result = this.getDataPermissionValue(this.collection.service.defaults.subjectData, permission)
        if (result == Tristate.UNDEFINED && this.opLevel >= Lantern.getGame().globalConfig.defaultOpPermissionLevel)
            result = Tristate.TRUE
        return result
    }

    override fun getOption(contexts: Set<Context>, key: String): Optional<String> {
        var result = super.getOption(contexts, key)
        if (!result.isPresent)
            result = getDataOptionValue(this.collection.defaults.subjectData, key)
        if (!result.isPresent)
            result = getDataOptionValue(this.collection.service.defaults.subjectData, key)
        return result
    }
}